File: Contests/CodeforcesPython/Educational Codeforces Round 132 Rated for Div. 2/C_Recover_an_RBS/C_Recover_an_RBS.py
import sys

# region fastio
input = lambda: sys.stdin.readline().rstrip()
sint = lambda: int(input())
mint = lambda: map(int, input().split())
ints = lambda: list(map(int, input().split()))
# print = lambda d: sys.stdout.write(str(d) + "\n")
# endregion fastio

def solve() -> None:
    s = input()
    cnt = qc = 0
    for c in s:
        if c == "(":
            cnt += 1
        elif c == ")":
            cnt -= 1
            if cnt < 0:
                # 当前)比(多，由于s必定是一个合法rbs，前面必须有一个?变成(
                cnt += 1
                qc -= 1
        else:
            if cnt > 0:
                qc += 1
            else:
                # 如果前面()一样多，当前?不可能变成)
                cnt += 1
        
        if cnt == 0 and qc == 1:
            # 如果前面()一样多，且存在一个?，则?不可能变成)
            cnt += 1
            qc -= 1
    
    print("YES" if abs(cnt) == qc or qc == 1 else "NO")

for _ in range(sint()):
    solve()

File: Contests/CodeforcesPython/Educational Codeforces Round 132 Rated for Div. 2/C_Recover_an_RBS/C_Recover_an_RBS.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
	// "sort"
)

func solve(_r io.Reader, _w io.Writer) {
	in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()

	var T int
	var s string
	for Fscan(in, &T); T > 0; T-- {
		Fscan(in, &s)

		cnt, qc := 0, 0
		for _, c := range s {
			if c == '(' {
				cnt++
			} else if c == ')' {
				cnt--
				if cnt < 0 {
					// 当前)比(多，由于s必定是一个合法rbs，前面必须有一个?变成(
					cnt++
					qc--
				}
			} else {
				if cnt > 0 {
					qc++
				} else {
					// 如果前面()一样多，当前?不可能变成)
					cnt++
				}
			}

			if cnt == 0 && qc == 1 {
				// 如果前面()一样多，且存在一个?，则?不可能变成)
				cnt++
				qc--
			}
		}

		if abs(cnt) == qc || qc == 1 {
			Fprintln(out, "YES")
		} else {
			Fprintln(out, "NO")
		}
	}
}

func main() { solve(os.Stdin, os.Stdout) }
func abs(a int) int {
	if a < 0 {
		return -a
	} else {
		return a
	}
}
